pkg/compact/dedup/replica.go

```diff
@@ -0,0 +1,258 @@
+}
+
+func (r *Replica) Group() string {
+	return replicaGroup(r.Labels)
+}
+
+func (r *Replica) isAggReplica() bool {
+	return r.Name == AggReplicaLabel
+}
+
+func (r *Replica) Resolution() (int64, error) {
+	if _, ok := r.Labels[ResolutionLabel]; !ok {
+		return -1, fmt.Errorf("no resolution found in replica group")
+	}
+	res, err := strconv.ParseInt(r.Labels[ResolutionLabel], 10, 64)
+	if err != nil {
+		return -1, err
+	}
+	return res, nil
+}
+
+type Replicas []*Replica
+
+// Group blocks by their global configs and resolution into different replicas.
+func NewReplicas(replicaLabelName string, blocks []*metadata.Meta) (Replicas, error) {
+	m := make(map[string]map[string][]*metadata.Meta, 0)
+	groupLabels := make(map[string]map[string]string)
+	for _, b := range blocks {
+		name, ok := b.Thanos.Labels[replicaLabelName]
+		if !ok {
+			return nil, errors.Errorf("not found replica label '%s' on block: %s", replicaLabelName, b.ULID.String())
+		}
+		labels := replicaLabels(replicaLabelName, b)
+		group := replicaGroup(labels)
+		groupLabels[group] = labels
+		if _, ok := m[group]; !ok {
+			m[group] = make(map[string][]*metadata.Meta, 0)
+		}
+		m[group][name] = append(m[group][name], b)
+	}
+	replicas := make(Replicas, 0)
+	for group, v := range m {
+		for name, blocks := range v {
+			replicas = append(replicas, NewReplica(name, groupLabels[group], blocks))
+		}
+	}
+	return replicas, nil
+}
+
+func replicaLabels(replicaLabelName string, b *metadata.Meta) map[string]string {
+	labels := make(map[string]string)
+	for k, v := range b.Thanos.Labels {
+		if k == replicaLabelName {
+			continue
+		}
+		labels[k] = v
+	}
+	labels[ResolutionLabel] = fmt.Sprint(b.Thanos.Downsample.Resolution)
+	return labels
+}
+
+func replicaGroup(labels map[string]string) string {
+	names := make([]string, 0, len(labels))
+	for k := range labels {
+		names = append(names, k)
+	}
+	sort.Slice(names, func(i, j int) bool {
+		return names[i] < names[j]
+	})
+	var b strings.Builder
+	b.WriteString("[")
+	for i, name := range names {
+		if i != 0 {
+			b.WriteString(",")
+		}
+		b.WriteString(fmt.Sprintf("%s=%s", name, labels[name]))
+	}
+	b.WriteString("]")
+	return b.String()
+}
+
+type ReplicaSyncer struct {
+	logger               log.Logger
+	metrics              *DedupMetrics
+	bkt                  objstore.Bucket
+	consistencyDelay     time.Duration
+	blockSyncConcurrency int
+
+	labelName string
+	mtx       sync.Mutex
+	blocksMtx sync.Mutex
+}
+
+func NewReplicaSyncer(logger log.Logger, metrics *DedupMetrics, bkt objstore.Bucket, labelName string,
+	consistencyDelay time.Duration, blockSyncConcurrency int) *ReplicaSyncer {
+	return &ReplicaSyncer{
+		logger:               logger,
+		metrics:              metrics,
+		bkt:                  bkt,
+		labelName:            labelName,
+		consistencyDelay:     consistencyDelay,
+		blockSyncConcurrency: blockSyncConcurrency,
+	}
+}
+
+func (s *ReplicaSyncer) Sync(ctx context.Context) (Replicas, error) {
```

Comment:
We already have this for compactor. Can we reuse somehow?

Reply:
same with above.